### ISSUE TITLE
AI Profiler not working well with .Net Core 3.0 and AKS

### ISSUE THREAD

**Issue by Eneuman** [microsoft/ApplicationInsights-Profiler-AspNetCore#48]
When using AI Profiler together with DEV-Spaces and AKS, I receive this error in my logs.

```
It was not possible to find any compatible framework version
The specified framework 'Microsoft.NETCore.App', version '2.0.0' was not found.
  - The following frameworks were found:
      3.0.0 at [/usr/share/dotnet/shared/Microsoft.NETCore.App]

You can resolve the problem by installing the specified framework and/or SDK.

The .NET Core frameworks can be found at:
  - https://aka.ms/dotnet-download
[41m[30mfail[39m[22m[49m: ServiceProfiler.EventPipe.Client.TraceUploader.TraceUploaderProxy[0]
      Trace upload failed.
```

I am creating my docker container using this build file:

```
FROM mcr.microsoft.com/dotnet/core/sdk:3.0
ARG BUILD_CONFIGURATION=Debug
ENV ASPNETCORE_ENVIRONMENT=Development
ENV DOTNET_USE_POLLING_FILE_WATCHER=true
ENV ASPNETCORE_URLS=http://+:80
EXPOSE 80

WORKDIR /src
COPY ["src/Test.Service/Test.Service.csproj", "src/Test.Service/"]

RUN dotnet restore "src/Test.Service/Test.Service.csproj"
COPY . .
WORKDIR "/src/src/Test.Service"
RUN dotnet build --no-restore "Test.Service.csproj" -c $BUILD_CONFIGURATION

RUN echo "exec dotnet run --no-build --no-launch-profile -c $BUILD_CONFIGURATION --" > /entrypoint.sh

ENTRYPOINT ["/bin/bash", "/entrypoint.sh"]

```

I find it strange that it says that it has found .Net Core 3.0 but it needs .Net Core 2.0.
I am using the latest beta-2 package.

Any idea on how to solve this?

**Comment by xiaomi7732** [comment]
@Eneuman, thanks for the report. I'll take a look.

**Comment by xiaomi7732** [comment]
@Eneuman, the issue is that the TraceUploader in the package works only when .NET Core 2.x exists, thus the failure.

While we are looking into the fix, here's a workaround:

1. Create a folder named 'Workaround' in your project folder and put the following zip file into the folder:
[TraceUpload.zip](https://github.com/microsoft/ApplicationInsights-Profiler-AspNetCore/files/3737011/TraceUpload.zip)

1. Update the docker file to overwrite the existing TraceUpload.zip:

```dockerfile
...
RUN dotnet publish -c Release -o out
...
# Work around - output the new uploader for .NET Core 3.0
COPY ./Workaround/TraceUpload.zip ./out/ServiceProfiler/TraceUpload.zip
...
```

My full dockerfile just as an FYI:
```dockerfile
FROM mcr.microsoft.com/dotnet/core/sdk:3.0 AS build-env
WORKDIR /app

# Copy csproj and restore as distinct layers
COPY *.csproj ./
RUN dotnet restore

# Copy everything else and build
COPY . ./
RUN dotnet publish -c Release -o out

# Work around - output the new uploader for .NET Core 3.0
COPY ./Workaround/TraceUpload.zip ./out/ServiceProfiler/TraceUpload.zip

# Build runtime image
FROM mcr.microsoft.com/dotnet/core/aspnet:3.0

WORKDIR /app
COPY --from=build-env /app/out .
ENTRYPOINT ["dotnet", "ServiceProfiler.EventPipe.DockerApp30.dll"]
```

Give it a try and feel free to let us know if there's more issue.

**Comment by cwoolum** [comment]
This works for me

**Comment by xiaomi7732** [comment]
@cwoolum Thanks for the comment. It is good to know that we have a working workaround.
The trace upload is a standalone app but targeting NetCoreApp2.0.
We are working on a fix to simship for both runtimes.

**Comment by cwoolum** [comment]
It would be great if the different versions could be hosted at a public endpoint and just added as a Dockerfile step. We actually add the profiler to one of our core libraries so our consuming apps actually need to have the zip manually added anyway.

We also use our own custom base docker images so downloading a file and adding it to the build as a step in the Dockerfile is much simpler than manually downloading the zip and copying it in.

**Comment by xiaomi7732** [comment]
@cwoolum Thanks for the feedback. Will it work for you if we post the binaries for the uploader in Github releases?

**Comment by cwoolum** [comment]
That would be the ideal solution for me!

**Comment by xiaomi7732** [comment]
The fix as been rolled out here: [2.0.0-beta1]().
Please follow the documentations here for migration:
* https://github.com/xiaomi7732/ApplicationInsights-Profiler-AspNetCore/tree/dev/saars/update-profiler-2.0/examples/QuickStart3_0
* https://github.com/xiaomi7732/ApplicationInsights-Profiler-AspNetCore/blob/dev/saars/update-profiler-2.0/docs/MigrateTo2_0.md

@cwoolum, please wait for a few days for us to release the binaries separately.

**Comment by xiaomi7732** [comment]
@cwoolum The binaries have been released here:
https://github.com/microsoft/ApplicationInsights-Profiler-AspNetCore/releases/tag/v2.0.0-beta1
